pallets/schemas/src/lib.rs

```diff
@@ -0,0 +1,143 @@
+use sp_runtime::{
+	DispatchError,
+	traits::{IdentifyAccount, Verify}
+};
+use sp_std::{convert::TryInto, vec::Vec};
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+pub use pallet::*;
+pub mod weights;
+pub use weights::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use common_primitives::schema::SchemaId;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		type WeightInfo: WeightInfo;
+
+		// TODO: may need to be moved into MSA and import traits into schemas pallet.
+		type Public: IdentifyAccount<AccountId = Self::AccountId>;
+		type Signature: Verify<Signer = Self::Public> + Member + Decode + Encode + TypeInfo;
+
+		// Maximum length of Schema field name
+		#[pallet::constant]
+		type MinSchemaSize: Get<u32>;
+
+		// Maximum length of Schema field name
+		#[pallet::constant]
+		type MaxSchemaSize: Get<u32>;
+
+		// Maximum number of schemas that can be registered
+		#[pallet::constant]
+		type MaxSchemaRegistrations: Get<SchemaId>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Emitted when a schemas is registered. [who, schemas id, block]
+		SchemaRegistered(T::AccountId, SchemaId, T::BlockNumber),
+	}
+
+	#[derive(PartialEq)] // for testing
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Schema is malformed
+		InvalidSchema,
+		/// The maximum number of schemas is stored in the database.
+		TooManySchemas,
+		/// The schemas exceeds the maximum length allowed
+		TooLongSchema,
+		/// The schemas is less than the minimum length allowed
+		TooShortSchema,
+		/// Schema does not exist
+		NoSuchSchema,
+		/// Error is converting to string
+		StringConversionError,
+		/// Error in Deserialization
+		DeserializationError,
+		/// Error in Serialization
+		SerializationError,
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	#[pallet::getter(fn schema_count)]
+	pub(super) type SchemaCount<T: Config> = StorageValue<_, SchemaId, ValueQuery>;
+
+	// storage for message schemas hashes
+	#[pallet::storage]
+	#[pallet::getter(fn get_schema)]
+	pub(super) type Schemas<T: Config> =
+		StorageMap<_, Twox64Concat, SchemaId, BoundedVec<u8, T::MaxSchemaSize>, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight((10_000, Pays::Yes))]
+		pub fn register_schema(origin: OriginFor<T>, schema: Vec<u8>) -> DispatchResult {
+			let sender = ensure_signed(origin.clone())?;
+
+			let cur_count = Self::schema_count();
+			ensure!(cur_count < T::MaxSchemaRegistrations::get(), <Error<T>>::TooManySchemas);
+			let schema_id = cur_count.checked_add(1).ok_or(<Error<T>>::TooManySchemas)?;
```

Comment:
This is an overflow error than not a TooManySchemas?

Reply:
I don't know why we'd draw the distinction though, either way it's too many.

Reply:
because these are two different errors:
```
ensure!(cur_count < T::MaxSchemaRegistrations::get(), <Error<T>>::TooManySchemas);

let schema_id = cur_count.checked_add(1).ok_or(<Error<T>>::TooManySchemas)?;
```

Reply:
After chatting we decided to keep both of these and change the second error to indicate an overflow even though (hopefully) that error is highly unlikely, it's a failsafe.